File: python/looker/rtl/user_session.py
"""UserSession to provid automatic authentication
"""
from typing import Dict, Optional
import urllib.parse

from looker.rtl import api_settings as st
from looker.rtl import auth_token as at
from looker.rtl import requests_transport as rtp
from looker.rtl import transport as tp
from looker.rtl import serialize as sr
from looker.rtl import model as ml


class UserSessionError(Exception):
    """Authentication problems.
    """


class UserSession():
    """UserSession to provid automatic authentication
    """

    def __init__(self,
                 settings: st.ApiSettings,
                 transport: Optional[tp.Transport] = None,
                 deserialize: Optional[sr.TDeserialize] = None):
        self._token: at.AuthToken = at.AuthToken()
        self.user_id: str = ''
        self.settings = settings
        self.transport = transport or rtp.RequestsTransport.configure(settings)
        self.deserialize = deserialize or sr.deserialize

    @property
    def is_authenticated(self) -> bool:
        """Determines if current token is active."""
        if not (self._token and self._token.access_token):
            return False
        return self._token.is_active

    @property
    def is_impersonating(self) -> bool:
        """Determines if user is impersonating another user."""
        return bool(self.user_id) and self.is_authenticated

    def get_token(self) -> at.AuthToken:
        """Returns an active token."""
        if not self.is_authenticated:
            self.login()
        return self._token

    def authenticate(self) -> Dict[str, str]:
        """Create a dictionary with the current token to be used as a header."""
        token = self.get_token()
        return {'Authorization': f'token {token.access_token}'}

    def login(self, user_id: Optional[str] = None) -> at.AuthToken:
        """Logs in a user or allows an already logged in user to sudo as another user."""
        if not self.is_authenticated:
            token = self._login(user_id)
            self._token = at.AuthToken(token)
        return self._token

    def _login(self, user_id: Optional[str]) -> ml.AccessToken:
        path = '/login'
        if (user_id and self.user_id != user_id):
            # We are switching user ids
            self._logout()
            self.user_id = user_id
            path += f'/{user_id}'

        self._reset()

        serialized = urllib.parse.urlencode({
            'client_id': self.settings.client_id,
            'client_secret': self.settings.client_secret
        }).encode('utf-8')
        response = self.transport.request(tp.HttpMethod.POST,
                                          path,
                                          body=serialized)
        if not response.ok:
            raise UserSessionError(str(response.value))
        token = self.deserialize(response.value, ml.AccessToken)
        if not isinstance(token, ml.AccessToken):
            raise UserSessionError(str(response.value))
        return token

    def logout(self) -> bool:
        """Logs out current active user.

        Returns true if the current user was logged in, False if not.
        """
        was_logged_in = False
        if self.is_authenticated:
            was_logged_in = True
            self._logout()
        return was_logged_in

    def _logout(self) -> bool:
        response = self.transport.request(tp.HttpMethod.DELETE,
                                          '/logout',
                                          authenticator=self.authenticate)

        if not response.ok:
            raise UserSessionError(str(response.value))

        if self.user_id:
            # Impersonated user was logged out, so set auth back to default
            self.user_id = ''
            self.login()
        else:
            self._reset()

        return True

    def _reset(self):
        self._token = at.AuthToken()

File: typescript/looker/rtl/userSession.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2019 Looker Data Sciences, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

import { IAccessToken, IError } from '../sdk/models'
import { IApiSettings } from './apiSettings'
import { IRequestInit, ITransport, SDKResponse, sdkError } from './transport'
import { AuthToken } from './authToken'
import { NodeTransport } from './nodeTransport'

// TODO support impersonation and reporting user id of logged in user?
export interface IUserSession {
  // Authentication token
  getToken(): Promise<IAccessToken>

  userId: string

  isImpersonating(): boolean

  isAuthenticated(): boolean

  authenticate(init: IRequestInit): Promise<IRequestInit>

  login(userId?: string): Promise<IAccessToken>

  logout(): Promise<boolean>

  settings: IApiSettings
  transport: ITransport
}


export class UserSession implements IUserSession {
  // TODO track both default auth token and user auth token, extract out token expiration logic to new class
  _token: AuthToken = new AuthToken()
  userId: string = ''
  transport: ITransport

  constructor(public settings: IApiSettings, transport?: ITransport) {
    this.settings = settings
    this.transport = transport || new NodeTransport(settings)
  }

  // Determines if the authentication token exists and has not expired
  isAuthenticated() {
    if (!(this._token && this._token.access_token)) return false
    return this._token.isActive()
  }

  isImpersonating() {
    return !!(this.userId && this.isAuthenticated())
  }

  async authenticate(init: IRequestInit) {
    const token = await this.getToken()
    if (token && token.access_token) init.headers.Authorization = `token ${token.access_token}`
    return init
  }

  async getToken() {
    if (!this.isAuthenticated()) {
      await this.login()
    }
    return this._token
  }

  // Reset the user session
  private reset() {
    this._token = new AuthToken()
  }

  private async ok<TSuccess, TError>(promise: Promise<SDKResponse<TSuccess, TError>>) {
    const result = await promise
    if (result.ok) {
      return result.value
    } else {
      throw sdkError(result as any)
    }
  }

  // internal login method
  private async _login(userId?: string) {
    if (userId && this.userId && this.userId !== userId) {
      // We're switching user ids
      await this.logout() // Logout the current user

    }
    this.reset()
    // authenticate client
    const result = await this.ok(this.transport.request<IAccessToken, IError>('POST', '/login',
      {client_id: this.settings.client_id, client_secret: this.settings.client_secret}))
    return result
  }

  async login(userId?: string) {
    if (!this.isAuthenticated()) {
      const token = await this._login(userId)
      this._token = new AuthToken(token)
      return token
    }
    return this._token
  }

  private async _logout() {
    // TODO verify this closure variable is required with a local block
    const token = this._token
    const promise = this.transport.request<string, IError>(
      'DELETE',
      '/logout',
      null,
      null,
      // ensure the auth token is included in the logout promise
      (init: IRequestInit) => {
        if (token.access_token) {
          init.headers.Authorization = `token ${token.access_token}`
        }
        return init
      }
    )

    await this.ok(promise)

    // If no error was thrown, logout was successful
    if (this.userId) {
      // User was logged out, so set auth back to default
      this.userId = ''
      await this.login()
    } else {
      // completely logged out
      this.reset()
    }
    return true
  }

  async logout() {
    let result = false
    if (this.isAuthenticated()) {
      result = await this._logout()
    }
    return result
  }

}
